### ISSUE TITLE
"git blame filename.c" does not work

### ISSUE THREAD

**Issue by kasal** [git-for-windows/git#105]
After upgrade from 1.9.5-preview-20150319 to 2.3.5.windows.8, "git blame filename.c" no longer works.
Every line is annotated as "Not Committed Yet", even though the working directory is clean.

"git blame HEAD -- filename.c" provides the correct output.

The obvious guess is that the CR chars have not been stripped, so all checksums are wrong.

My setup looks like this:
git config core.autocrlf = False
file .gitattributes contains
.c     text

**Comment by dscho** [comment]
Hmm. Purportedly https://github.com/git/git/commit/4d4813a52f3722854a54bab046f4abfec13ef6ae fixed this, though...

**Comment by tboegi** [comment]
Not sure if 4d481 fixed it: It changes the behaviour for core.autocrlf=input.
(My very feeling is, that we need something similar for the attributes as well,
 but no one has reported this issue yet)

Stepan,
(I assume that your .gitattributes uses *.c text)

Which line endings does the file have in the working tree ?
Which line endings in the repo ?
is there a way to make this reproducable ?
Can you do a bisect ?

**Comment by kasal** [comment]
Hi tboegi and dscho,
the file has CRLF in the working tree and LF in the repo.
Yes, I can reproduce with my 2.3.5.windows.8 install: attached below please find test cases in a form suitable for addition to git's test suite.

dscho wrote:
> Hmm. Purportedly git@4d4813a52f3722854a54bab046f4abfec13ef6ae fixed this, though...

Close.  Actually, the bug was _introduced_ by this commit and shall be fixed by reverting it.
Thank you for pointing to the right spot.

Commit git@4d4813a52f3722854a54bab046f4abfec13ef6ae deals with a case when the repo contains files with CRLF. This setup is invalid.  (It can work in certain cases, but keeping all files set as binary is the very minimum.)

After changing core.autocrlf, one has to normalize the line endings in the repo, see e.g. [this description](https://help.github.com/articles/dealing-with-line-endings/#refreshing-a-repository-after-changing-line-endings).

The commit should be reverted, including the incorrect test case; another test cases shall be added instead.

I would be very glad if any of you could volunteer to submit this upstream.

```sh
test_expect_success 'blaming files with CRLF newlines' '
	git config core.autocrlf true &&
	printf "testcase\r\n" >crlffile &&
	git add crlffile &&
	git commit -m testcase &&
	git blame crlffile >actual &&
	grep "A U Thor" actual
'
test_expect_success 'blaming files with CRLF newlines' '
	git config core.autocrlf false &&
	printf ".gitattributes text\r\n" >.gitattributes &&
	git add .gitattributes &&
	git commit -m testcase &&
	git blame .gitattributes >actual &&
	grep "A U Thor" actual
'
```

**Comment by tboegi** [comment]
Agreed, this commit should be reverted.

I can volunteer to add a test case for git blame to t0027,
which systematically checks all combinations.
 (at some time in the future)

**Comment by kasal** [comment]
Let me look at the possible combinations:
- CRLF in repo is not supported,
- if both repo and working tree are LF, no specific test is needed.
Only one combination remains: repo LF, work tree CRLF
IMHO the two cases mentioned above are representative enough.

**Comment by dscho** [comment]
@kasal Feel free to cherry-pick [this commit](https://github.com/git/git/commit/4bf256d67a85bed1e175ecc2706322eafe4489ca), verify that it fixes the problem you described and open a Pull Request. I would be delighted to merge it.

**Comment by dscho** [comment]
This was fixed by merging #138.
